### DISCUSSION TITLE
Issues with uploading images to my vercel app

### DISCUSSION THREAD

**Discussion by smarbo** [vercel/community#3582] [Help] [answered]
### Summary

Hi there, I have been having issues with uploading images to my vercel app, munchbox. its a recipe sharing app and each recipe has to have an image. Locally hosted, the image uploading works fine. The images are uploaded to /.next/server/pages/api/uploads/recipes/{the image name}.jpg, and work fine on local devlopment server. but when i try on the vercel app, i get this error:
{
    "error": {
        "errno": -2,
        "code": "ENOENT",
        "syscall": "open",
        "path": "/var/task/.next/server/pages/api/uploads/recipes/1691857581665.jpg"
    }
}

The code for the project is at [the github repo](https://github.com/smarbo/munchbox)
If anyone knows how to fix this it would be great.
also dont tell me to use amazon s3 because it doesn't work for me

### Example

https://munchbox.vercel.app/recipes/new

### Steps to Reproduce

The api endpoint code for the image uploading thing:
```javascript
// Import necessary packages
require("dotenv").config();
import ApiHandler from "@/servercomponents/apihandler";
import path from "path";
import Recipe from "@/models/recipe";
require("@/servercomponents/db");
const uploadPath = path.join(__dirname, "uploads", "recipes");
const fs = require("fs/promises");

async function initUploadDirs() {
    try {
        await fs.access(uploadPath);
    } catch (error) {
        if (error.code === "ENOENT") {
            await fs.mkdir(uploadPath, { recursive: true });
        } else {
            throw error;
        }
    }
}

initUploadDirs();

// Export the API route handler
export default (req, res) => {
    new ApiHandler({
        post: createRecipe,
        get: allRecipes,
        authKeys: [...JSON.parse(process.env["AUTH_KEY"])],
    }).handleRequest(req, res);
};

export const config = {
    api: {
        bodyParser: {
            sizeLimit: "8mb",
        },
    },
};

// Function to create a recipe for POST requests
async function createRecipe(req, res) {
    const parsedBody = JSON.parse(req.body);
    const { title, time, creator, ingredients, content, image } = parsedBody;
    const recipeId = Date.now();
    if (!title || !time || !creator || !ingredients || !content) {
        return res.status(400).json( {
            error: "Invalid Data",
            reqBody: [title, time, creator, ingredients, content],
        });
    }
    try {
        const base64Data = image.split(",")[1];
        const imageBuffer = Buffer.from(base64Data, "base64");
        const filePath = path.join(uploadPath, `${recipeId}.jpg`);
        console.log(filePath);
        await fs.writeFile(filePath, imageBuffer);
        await Recipe.create({
            title,
            time,
            creator,
            ingredients,
            content,
            recipeId,
        });
        console.log("Success");
        return res.status(200).json({ message: "Success!" });
    } catch (err) {
        console.log(err);
        return res.status(500).json({ error: err });
    }
}

// Function for GET allRecipes endpoint
function allRecipes(req, res) {
    try {
        const recipes = Recipe.find({});
        return res.status(200).json({ data: recipes });
    } catch (err) {
        console.log(err);
        return res.status(500).json({ error: err });
    }
}

```

**Comment by amyegan** [comment, marked as answer]
Hey @smarbo. Because of the dynamic nature of Vercel's serverless environment, you won't be able to write additional files outside of the normal build & deploy steps. You can [use temporary storage](https://vercel.com/guides/how-can-i-use-files-in-serverless-functions#using-temporary-storage) to upload a file in a serverless function before writing it to a more permanent storage location. The example code does use S3, but you can use whatever storage location works best for you. 

This guide on using files in serverless functions has more info: https://vercel.com/guides/how-can-i-use-files-in-serverless-functions

**Reply by smarbo** [reply]
Thanks, I've already implemented cloudinary in my newest version and it works fine even on the vercel deployment.

**Reply by vnevermore** [reply]
could you share with the communitu how you solved your issue?

**Reply by smarbo** [reply]
Yeah, I used a cloud service called cloudinary and it's free you just need to make an account and it should have the API reference on the website with a quick start guide 

. I had an issue where I couldn't download the files locally to my vercel deployment so I had to somehow stream it directly from the user's upload. Cloudinary has a feature in their API here you can stream an image buffer to the server instead of downloading the file on the server and then sending it. You can look it up on their API reference if you need this too. 

You can also just send a file that is downloaded on the server but if you're using vercel deployment I don't know how you would do that.

For reading the images, I just stored the 'secure url' (URL to the image on clodyinary's server) in each object in the database as a string, and it worked fine. 

If you need any more help just ask me or if you want to somehow try and understand my code it's in my GitHub repo [munchbox](Https://GitHub.com/smarbo/munchbox) in the file pages/API/recipes/index.js in the POST function

**Reply by alexanderdomrachev** [reply]
I have a "400 Bad Request" for my external images with the response "url" parameter not allowed (in production).  On localhost everything works fine. RemotePatterns in next.config.js for external domains are prescribed. I think the problem might be in nginx, but I don't have a solution yet. Images from /public in production are loaded ok, external images are not.
